proposals/0545-deterministic-operator-parsing.md

```diff
@@ -0,0 +1,157 @@
+e.g. Ormolu, apply-refact, retrie, hlint.
+
+## Motivation
+
+Haskell supports custom operators and allows you to specify custom precedences and
+associativities for the custom operators.
+However, this means you can _not_ parse Haskell code without knowing how the operators
+were defined, e.g. `a $$$ b $$$ c` could mean either `(a $$$ b) $$$ c` or `a $$$ (b $$$ c)`.
+
+This means formatters like Ormolu need to keep a database of all common operators,
+and support specifying custom operators in the configuration file, rather than Just Working™.
+See e.g. https://github.com/tweag/ormolu/issues/927
+
+Nim already does a similar thing: https://nim-lang.org/docs/manual.html#syntax-associativity
+
+## Proposed Change Specification
+
+When Haskell is parsed with this `-XDeterministicOperatorParsing` enabled,
+all associativities and precedences will be determined from the name, and name only.
+
+Operators from `base` (as of 4.18.0.0) are special-cased and retain their current behaviour.
+New operators in `base` are subject to the same rules as for custom operators.
+
+### Associativity
+
+Associativity for custom operators will be determined from the first symbol in the name.
+The table for which associativity is assigned is listed in the appendix.
+
+### Predence
+
+Determining the relative precedence of two custom operators in an expression
+`x A y B c` is done via "comparing" them according to the order in the table,
+such that if A's first symbol is higher than B's first symbol, then A is higher,
+if they're equal, then look at the next symbol, or in the reverse case, B is higher.
+
+When determining the relative precedence of a custom operator and a special-cased one,
+the same algorithm as above is used, but the special-cased operator is considered to consist
+of an infinite repetition of the same pseudo-symbol that has as precedence the operators precedence.
+
+That is, if the first symbol in the custom operator has a precedence different from that of
+the special-cased one, then the relative order of those two precedences are used,
+but if the first symbol has a precedence _equal_ to the one of the special-cased operator,
+then the next symbol of the operator is considered.
+
+## Examples
+
+`a +| b *| c` is equivalent to `a +| (b *| c)`.
+
+`a +| b -> c` is equivalent to `(a +| b) -> c`.
+
+`a $$$ b $$$ c` is equivalent to `a $$$ (b $$$ c)`.
+
+`a ~~~ b ~~~ c` is invalid.
+
+`a +* b + c` is equivalent to `(a +* b) + c`.
+
+## Effect and Interactions
+
+Given _only_ the precedences and associativities of the operators in `base` as of now,
+any Haskell file can be successfully parsed, formatted, and manipulated,
+if `-XDeterministicOperatorParsing` is enabled.
+
+Formatters would be able to take advantage of this fact if told to use it,
+or if it's enabled via a language pragma.
+
+## Costs and Drawbacks
+
+Adding a new language extension adds complexity. In addition, the
+same code can mean _different_ things depending on whether the extension
```

Comment:
This sounds like major trouble. Your goal is to help tooling, but now tooling needs to deal with modules _with_ and _without_ the extension, producing different results! Let me suggest an alternative.

Your ultimate goal seems to be able to determine the fixity of an operator locally, without looking at other modules. So let's keep fixity declarations, but require specifying them when importing operators. Instead of `import M (foo, bar, (+|))`, ask the user to write:

```
import M (foo, bar, infixl 6 +|)
```

And emit a warning (that could be made into an error with `-Werror`) for any operator occurrence whose fixity is not specified in the import list.

Make an exception for `Prelude`.

Reply:
Alternatively, now that I see a similar suggestion from @treeowl, this does not need to be part of the import list.

Just permit duplicate fixity declarations at the top-level:

```
import M (foo, bar, (+|))

infixl 6 +|
```

No need to modify the import list syntax, no need to introduce a pragma. Just list the fixities of the operators used in the current module, and the compiler will check that those fixities agree with the ones specified when the operator was defined.

Reply:
> Alternatively, now that I see a similar suggestion [...] this does not need to be part of the import list.

Just to add that I found your first idea much better. Making it part of the import list intuitively conveys the idea that the fixity has to be declared in the module from which it is imported. Just repeating the fixity declaration at the toplevel intuitively conveys that the "ground source of truth" about the fixity of the operator is given at that place.

Reply:
I agree with @BinderDavid, for whatever it's worth.

Reply:
On the other hand: is there anything wrong with allowing an operator to be imported with an intentionally different fixity? Could we mark whether the fixity should merely be checked or actually overridden? That would leave open the question of fixity for re-exported operators.

Reply:
Allowing a different fixity would be very confusing.